src/server.rs

```diff
@@ -42,48 +42,130 @@ pub trait Server {
     #[rpc(name = "swap")]
     fn swap(&self, swap: SwapReq) -> Result<Value>;
 
-    // milestone 3:
+    // milestone 3: Used by mwixnet coinswap servers to communicate with each other
     // fn derive_outputs(&self, entries: Vec<Onion>) -> Result<Value>;
     // fn derive_kernel(&self, tx: Tx) -> Result<Value>;
 }
 
+#[derive(Clone)]
 pub struct ServerImpl {
-    server_key: SecretKey,
+    server_config: ServerConfig,
+
+    wallet: Wallet,
+
+    node: GrinNode,
 }
 
 impl ServerImpl {
-    pub fn new(server_key: SecretKey) -> Self {
-        ServerImpl { server_key }
+    pub fn new(server_config: ServerConfig, wallet: Wallet, node: GrinNode) -> Self {
+        ServerImpl { server_config, wallet, node }
+    }
+
+    /// The fee base to use. For now, just using the default.
+    fn get_fee_base(&self) -> u64 {
+        DEFAULT_ACCEPT_FEE_BASE
+    }
+
+    /// Minimum fee to perform a swap.
+    /// Requires enough fee for the mwixnet server's kernel, 1 input and its output to swap.
+    pub fn get_minimum_swap_fee(&self) -> u64 {
+        TransactionBody::weight_by_iok(1, 1, 1) * self.get_fee_base()
+    }
+
+    /// Iterate through all saved submissions, filter out any inputs that are no longer spendable,
+    /// and assemble the coinswap transaction, posting the transaction to the configured node.
+    /// 
+    /// Currently only a single mix node is used. Milestone 3 will include support for multiple mix nodes.
+    pub fn execute_round(&self) -> crate::error::Result<()> {
+        let locked_state = SERVER_STATE.lock().unwrap();
+        let next_block_height = self.node.get_chain_height()? + 1;
+
+        let spendable : Vec<Submission> = locked_state
+            .iter()
+            .filter(|s| self.node.is_spendable(&s.input.commit, next_block_height).unwrap_or(false))
+            .cloned()
+            .collect();
+
+        let total_fee : u64 = spendable
+            .iter()
+            .enumerate()
+            .map(|(_, s)| s.fee)
+            .sum();
+        
+        let inputs : Vec<Input> = spendable
+            .iter()
+            .enumerate()
+            .map(|(_, s)| s.input)
+            .collect();
+        
+        let outputs : Vec<Output> = spendable
+            .iter()
+            .enumerate()
+            .filter_map(|(_, s)| s.output)
+            .collect();

```

Comment:
Where is verification of rangeproofs

Reply:
oops.😳 I had a plan for where to add rangeproof verification for multiple servers in milestone 3, but I overlooked that step for the single server impl. I'll add an additional filter on line 86, just after the `is_spendable` filter.